fdbclient/include/fdbclient/MetaclusterManagement.actor.h

```diff
@@ -1415,13 +1415,33 @@ struct DeleteTenantImpl {
 				// before we could mark it as removed and we are trying to remove the "new" name.
 				// We can proceed as normal, but treat it as a standalone remove instead of paired
 				if (!pairEntry.present() || pairEntry.get().id != self->tenantId) {
-					self->pairName.reset();
-					return Void();
+					// Another edge case when chaining renames together and a delete comes in between:
+					// Rename A->B. Mark both as renaming
+					// Delete B. It sees pair A in getAssignedLocation
+					// Rename A->B completes. Entry A is removed from management cluster
+					// Rename B->C. Mark both as renaming (before we mark B as Removing here)
+					// Reach here and see A is no longer present. Instead of resetting the pair
+					// we need to update our pair here to C, so we don't have inconsistent state
```

Comment:
If I'm understanding this correctly, there should be two possible outcomes from this:

1. If rename B->C is recorded in the management cluster before updating state in the management cluster for the delete, then delete A should fail.
2. If rename B->C occurs after updating the state in the management cluster for the delete, then the rename should fail.

Reply:
One way to think about these operations that occur over multiple transactions is that there is one transaction (usually the first on the management cluster that changes state) that is used to declare that the change will be taking place, and it sort of acts as a lock on the tenant metadata being modified until the final management transaction resolves.

The exception to this is that removal is allowed to happen to a tenant in a transitory phase, but that would then block anything else from happening to the tenant afterward. Also, any operation is allowed to be repeated while in flight (so you could issue rename A->B again while A->B is in progress in order to resume the rename).

In this case, if we've started the removal of A while the rename A->B was in progress (i.e. if we recorded the removal in the management cluster), then the final A->B transaction should not be allowed to complete and it would not be possible to rename B->C because the remove has locked both A and B.

If we've started a rename on B->C, then that must mean that A->B has completed given that the rename on A->B would otherwise have locked B from being renamed. If the A->B rename has completed, deleting A should have no impact on B or C; it is either not present and should throw an error or has been recreated and should be unassociated with the former renamed tenant.

Reply:
So in this scenario, we are deleting B, not A. Technically, we are in scenario 2, and the rename does actually fail, because it sees the state as removing in the final management transaction. I don't think scenario 1 can actually execute properly because like you said:
> removal is allowed to happen to a tenant in a transitory phase

Thus there is no criteria for us to make the delete fail.
 The problem here is that without this change, the Delete on B thinks its pair is still A, and will end up leaving a temporary tenant C in the management cluster.

Reply:
I guess the other thing that we should do is to not set or use the pair ID at all until the second management transaction. We optionally could use it in the first data transaction, but it shouldn't be necessary given that the tenants have the same ID (and therefore the same prefix).

Regardless, we should unconditionally read and set the pair name in the second management transaction when we are actually recording the removal.

Reply:
For some reason I remember us deciding to allow deletes on both renaming states in the end, but if we can just flat out reject the `RENAMING_TO` case then I think we can have fewer checks for edge cases. I will also try to move more of the state changes and checks to the second management transaction

Reply:
We discussed that it was ok to allow deleting the pair once the removal had already started (i.e. both sides were in the removing phase, so we couldn't tell which was the source anymore).